### PR TITLE
feat: add cursor worker pre-trust helper

### PR DESCRIPTION
**Pull request by baksohyeon**
## Problem

On 2026-08-03, Cursor Agent workers dispatched into fresh Orca worktrees stopped on a workspace-trust prompt before doing any contract work. Trust persistence was measured from known-trusted `.orca/worktrees/mogui-ADE-orchestrator/*` worktrees and controlled untrusted launches: Cursor stores trust in `~/.cursor/projects/<project-key>/.workspace-trusted` as JSON (`workspacePath`, `trustedAt`), with the key form derived from the absolute path by mapping `/` → `-` and removing `.`. Without a pre-trust helper parallel to `scripts/codex-worker-pretrust`, every Cursor attach re-hit that interactive gate. After the initial PR opened, 8 of 16 review threads remained unresolved (measured 2026-08-04) and the PR body did not follow the repository's four-section narrative template.

## Why this approach

Alternative not taken: rely on Cursor's launch `--trust` / `--force` flags alone. Those still require the master to re-measure CLI help at every dispatch and do not pre-seed the same on-disk marker Cursor itself writes after a human trusts a workspace; a worktree-path helper that writes the measured marker is the same shape operators already use for Codex and survives relaunches. Constraint that forced this shape: the marker location and key form are host-measured, not documented by Cursor as a stable public API, so the script stays small, idempotent, and loud when it cannot parse or safely write.

## What this changes

- Adds `scripts/cursor-worker-pretrust` (absolute worktree path; optional `--projects-dir` / `CURSOR_PROJECTS_DIR`) that writes/validates `.workspace-trusted` JSON and prints the same summary style as the Codex helper.
- Hardens path handling: trailing-slash symlink rejection, non-regular marker refusal, no-follow directory FDs + atomic replace for writes, Python 3.6+ capability probe with versioned `python3.14`…`python3.6` candidates.
- Adds `tests/test_cursor_pretrust.py` for add/idempotence/malformed/mismatch/symlink/interpreter/non-regular cases.
- Documents the command in `docs/public/reference.md` (help-derived options only) and Cursor pre-trust in `master-ops/docs/charter/04-worker-routing-review.md` / `05-dispatch-gate.md` plus both changelogs.
- Records follow-up measurement that `cursor-agent 2026.07.23-e383d2b` showed no second startup trust gate for project hooks after pre-trust.
- Intentionally out of scope: changing Codex pre-trust, inventing a second Cursor trust store, raising `master-ops/TEMPLATE-VERSION` (no template release cut here), and merging this PR.

## Expected effect

A master can run `scripts/cursor-worker-pretrust <abs-worktree>` before attach and observe exit 0 with `Summary: N added|updated|already trusted` (not `skipped`). A second run on the same path reports `already trusted`. Missing paths exit 2; symlink destinations and non-regular markers exit 2 with explicit errors. Verify with:

```bash
scripts/cursor-worker-pretrust /absolute/path/to/worktree
PYTHONPATH=src python3.12 -m pytest tests/test_cursor_pretrust.py -q
```

Public command discovery: `docs/public/reference.md` row for `cursor-worker-pretrust`. Runbook: `master-ops/docs/charter/04-worker-routing-review.md` (Worker Launch Approval Posture).

## Testing

- [x] `PYTHONPATH=src python3.12 -m pytest tests -q` — `466 passed, 13 subtests passed` (host `python3` 3.9 lacks some suite deps; 3.12 is the repository interpreter used here).
- [x] A test that fails without this change — `tests/test_cursor_pretrust.py` (16 tests) covers fresh add, idempotence, malformed JSON, mismatch, symlink rejection (including trailing-slash projects-dir), non-regular marker, versioned interpreter discovery including python3.10-only PATH, minimum capability probe, and skip-without-write.
- [x] New files staged before running any scan. `./scripts/redaction-scan.sh` — `OK — 0 findings (mode=tracked, files=177, commit-messages=not-scanned, org-rules=0)`.

## Review

- Automated review (CodeRabbit + cubic) filed 8 unresolved threads after main sync: TOCTOU write race, trailing-slash symlink bypass, non-regular marker crash, macOS tmp_path symlink ancestors in tests, missing python3.10 candidate, weak interpreter probe, MD001 heading skip, and public-reference measurement digression.
- Each finding was reproduced or measured where applicable, fixed in `0dcfef6`, replied to individually, and resolved. No PR-level blanket resolve.
- Heading fix is pure hierarchy (MD001). Measurement note moved out of `docs/public/reference.md` while retained in runbook/changelogs.

## Changelog

- Root `CHANGELOG.md` `[Unreleased]`: new script, measured trust storage, and hardening follow-up (no-follow writes, trailing-slash symlink rejection, non-regular guard, broader Python probe).
- `master-ops/CHANGELOG.md` `[Unreleased]`: Cursor pre-trust runbook guidance and second-gate measurement note.

## Template impact

Touches `master-ops/` docs only (charter worker-routing heading + pre-trust guidance already present; heading level corrected). Existing installations that want this behavior should copy the Cursor pre-trust paragraph in the worker-routing / dispatch-gate charter slices and the corresponding changelog note into their operations copy. `master-ops/TEMPLATE-VERSION` is unchanged in this PR because no template release is being cut here.

## Notes

- Head after stewardship: `0dcfef6` (`fix: harden cursor-worker-pretrust path and interpreter probes`).
- Provenance: trust storage location and key form measured 2026-08-03; second-gate absence measured on `cursor-agent 2026.07.23-e383d2b`.
- Platform: hardening uses POSIX `O_NOFOLLOW` / directory FDs; hosts without `O_NOFOLLOW` refuse to write rather than degrade to follow-symlink opens.

### CODE REVIEW

File: master-ops/docs/MASTER-OPERATIONS.md [hunk 1]
```diff
@@ -124,13 +124,14 @@ The installation-specific worker tier mapping lives in `model-tier-policy.json`,
...
 
 MEASURED examples from 2026-08-02:
 
 - Grok: `--always-approve`.
 - Claude Code: `--dangerously-skip-permissions`.
 - Cursor Agent: `--force` (also exposed as `--yolo`) to force-allow commands unless explicitly denied, plus `--trust` to trust the current workspace without prompting.
+- Cursor Agent pre-trust: run `scripts/cursor-worker-pretrust <worktree-path>` before attach and confirm the summary is not `skipped`; this uses Cursor Agent's measured `.workspace-trusted` storage instead of relying on `--trust` at worker launch.
 - Codex: run `scripts/codex-worker-pretrust <worktree-path>` and retain the account-wide hooks trust posture; Codex uses this pre-trust path instead of a launch approval flag.
```
[the remark this answers is not in the document]
Comment (baksohyeon) [reply master-ops/docs/MASTER-OPERATIONS.md]: Fixed in 8a60950. I split the Cursor pre-trust item into its own "MEASURED addition from 2026-08-03" block so the date now matches the measured evidence and changelog entries.

File: scripts/cursor-worker-pretrust [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    break
+  fi
+done
+if [[ -z "$python_bin" ]]; then
+  skip_message='SKIP no Python interpreter with json support found (tried python3, python); trust marker edits are verified by parsing JSON, so nothing was changed and dispatched Cursor workers will see a trust prompt'
+  printf '%s\n' "$skip_message"
+  printf '%s\n' "$skip_message" >&2
+  printf 'Summary: skipped — 0 added, 0 updated, 0 already trusted\n'
+  exit 0
+fi
```
[the remark this answers is not in the document]
Comment (baksohyeon) [reply scripts/cursor-worker-pretrust:100]: Fixed in 8a60950. The interpreter probe now requires Python 3 (`sys.version_info[0] >= 3`) before selecting a candidate, so a Python 2 `python` fallback can no longer reach the embedded script path.

File: scripts/cursor-worker-pretrust [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    except json.JSONDecodeError as error:
+        abort(f"cannot parse trust marker JSON: {error}")
+    if not isinstance(document, dict):
+        abort("trust marker is not a JSON object")
+    return document
+
+
+if not trust_path.exists():
+    trust_path.parent.mkdir(parents=True, exist_ok=True)
+    trust_path.write_text(
```
[the remark this answers is not in the document]
Comment (baksohyeon) [reply scripts/cursor-worker-pretrust]: Addressed in e86d0bb. The script now rejects symlinked path components for `projects_dir`, the derived project directory, and `.workspace-trusted` before any read/write, preventing writes through symlink redirects.

File: scripts/cursor-worker-pretrust [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+project_key=${trimmed_path//\//-}
+project_key=${project_key//./}
+if [[ -z "$project_key" ]]; then
+  printf 'ERROR could not derive a project key from path: %s\n' "$worktree_path" >&2
+  exit 2
+fi
+trust_file="$projects_dir/$project_key/.workspace-trusted"
+
+python_bin=""
+for python_candidate in python3 python; do
```
[the remark this answers is not in the document]
Comment (baksohyeon) [reply scripts/cursor-worker-pretrust]: Addressed in e86d0bb. Interpreter probing now searches `python3`, `python3.14`, `python3.13`, `python3.12`, `python3.11`, then `python`, matching the Codex helper style while still enforcing Python 3 capability.

File: master-ops/docs/MASTER-OPERATIONS.md [hunk 5]
```diff
@@ -174,6 +178,8 @@ The verdict is graded. No declared model, or a probe that returns nothing, warns
 
 Before attaching a Codex worker, run `scripts/codex-worker-pretrust <worktree-path>` and confirm the summary is not `skipped`, so startup does not block on the trust prompt.
 
+Before attaching a Cursor worker, run `scripts/cursor-worker-pretrust <worktree-path>` and confirm the summary is not `skipped`, so startup does not block on the trust prompt.
```
[the remark this answers is not in the document]
Comment (baksohyeon) [reply master-ops/docs/MASTER-OPERATIONS.md]: Addressed in e86d0bb. The runbook now requires exit status 0 and a summary containing `added`, `updated`, or `already trusted` (not `skipped`) for Cursor pre-trust attach readiness.

File: master-ops/docs/MASTER-OPERATIONS.md [hunk 6]
```diff
@@ -133,6 +133,10 @@ MEASURED examples from 2026-08-02:
 - Cursor Agent: `--force` (also exposed as `--yolo`) to force-allow commands unless explicitly denied, plus `--trust` to trust the current workspace without prompting.
 - Codex: run `scripts/codex-worker-pretrust <worktree-path>` and retain the account-wide hooks trust posture; Codex uses this pre-trust path instead of a launch approval flag.
 
+MEASURED addition from 2026-08-03:
+
+- Cursor Agent pre-trust: run `scripts/cursor-worker-pretrust <worktree-path>` before attach and confirm the summary is not `skipped`; this uses Cursor Agent's measured `.workspace-trusted` storage instead of relying on `--trust` at worker launch.
```
[the remark this answers is not in the document]
Comment (baksohyeon) [reply master-ops/docs/MASTER-OPERATIONS.md]: Addressed earlier in 8a60950 by splitting this into a separate "MEASURED addition from 2026-08-03" section.

File: tests/test_cursor_pretrust.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    shim_bin = tmp_path / "bin"
+    shim_bin.mkdir()
+    (shim_bin / "bash").symlink_to("/bin/bash")
+    env = os.environ.copy()
+    env["PATH"] = str(shim_bin)
+
+    result = run_pretrust(str(workspace), projects_dir, env=env)
+
+    assert result.returncode == 0
+    assert "SKIP no Python 3 interpreter with json support found" in result.stdout
```
[the remark this answers is not in the document]
Comment (baksohyeon) [reply tests/test_cursor_pretrust.py]: Addressed in 572c554. Added direct regression coverage for the Python-2-fallback scenario (`test_python_candidate_that_is_not_python3_is_rejected`) to prove the probe rejects non-Python-3 `python` and safely skips without writing a marker.

File: scripts/cursor-worker-pretrust [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+project_key=${trimmed_path//\//-}
+project_key=${project_key//./}
+if [[ -z "$project_key" ]]; then
+  printf 'ERROR could not derive a project key from path: %s\n' "$worktree_path" >&2
+  exit 2
+fi
+trust_file="$projects_dir/$project_key/.workspace-trusted"
+project_dir="$projects_dir/$project_key"
+
+for symlink_candidate in "$projects_dir" "$project_dir" "$trust_file"; do
```
[the remark this answers is not in the document]
Comment (baksohyeon) [reply scripts/cursor-worker-pretrust]: Addressed in b38f01f. The symlink guard now walks ancestor components without external tools, so symlinks anywhere along the `projects_dir`/project path are rejected before any trust marker read/write.

**Comment by baksohyeon** [comment]
@cubic-dev-ai please re-review latest commit b38f01f; all previously reported threads are resolved.

File: scripts/cursor-worker-pretrust [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+if workspace_in_file is None:
+    updated["workspacePath"] = workspace_path
+if not isinstance(trusted_at, str) or not trusted_at:
+    updated["trustedAt"] = now_iso()
+
+encoded = json.dumps(updated, indent=2) + "\n"
+if trust_path.read_text(encoding="utf-8") == encoded:
+    print("already")
+else:
+    trust_path.write_text(encoded, encoding="utf-8")
```
[the remark this answers is not in the document]
Comment (baksohyeon) [reply scripts/cursor-worker-pretrust]: Thanks — valid race concern. The current symlink check is preflight-only, so a TOCTOU window remains between check and write. I have merged main and retriggered checks; I will follow with a hardening patch that uses race-resistant open/write semantics (no-follow + atomic update) and add a regression test.

File: scripts/cursor-worker-pretrust [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+project_key=${project_key//./}
+if [[ -z "$project_key" ]]; then
+  printf 'ERROR could not derive a project key from path: %s\n' "$worktree_path" >&2
+  exit 2
+fi
+trust_file="$projects_dir/$project_key/.workspace-trusted"
+project_dir="$projects_dir/$project_key"
+
+check_no_symlink_ancestors() {
+  local candidate=$1
```
[the remark this answers is not in the document]
Comment (baksohyeon) [reply scripts/cursor-worker-pretrust:52]: Good catch. A trailing slash should be normalized before each symlink test so the leaf component itself is always checked. I have merged main and retriggered checks; I will include this normalization fix with a focused regression test for symlinked --projects-dir values ending in `/`.

File: scripts/cursor-worker-pretrust [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+            },
+            indent=2,
+        )
+        + "\n",
+        encoding="utf-8",
+    )
+    print("added")
+    raise SystemExit(0)
+
+document = parse_document(trust_path.read_bytes())
```
[the remark this answers is not in the document]
Comment (baksohyeon) [reply scripts/cursor-worker-pretrust]: Confirmed. Existing non-regular marker paths should fail with a stable, explicit error path before read/parse. I have merged main and retriggered checks; I will add a regular-file guard and test coverage for non-regular `.workspace-trusted` entries.

File: tests/test_cursor_pretrust.py [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+    assert trusted_file(
+        projects_dir,
+        "/tmp/.orca/worktrees/repo/feature.with.dot",
+    ) == Path(
+        "/tmp/projects-root/tmp-orca-worktrees-repo-featurewithdot/.workspace-trusted"
+    )
+
+
+def test_fresh_add_writes_dot_stripped_project_key(tmp_path: Path) -> None:
+    projects_dir = tmp_path / "projects"
```
[the remark this answers is not in the document]
Comment (baksohyeon) [reply tests/test_cursor_pretrust.py:49]: Thanks for flagging this edge case. I re-ran the suite on macOS and tests currently pass, but your point about fixture path normalization is valid for portability and clarity. I have merged main and retriggered checks; I will add explicit resolved-path handling in test fixtures to avoid symlink-ancestor ambiguity.

File: scripts/cursor-worker-pretrust [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+    fi
+  done
+}
+
+check_no_symlink_ancestors "$projects_dir"
+check_no_symlink_ancestors "$project_dir"
+check_no_symlink_ancestors "$trust_file"
+
+python_bin=""
+for python_candidate in python3 python3.14 python3.13 python3.12 python3.11 python; do
```
[the remark this answers is not in the document]
Comment (baksohyeon) [reply scripts/cursor-worker-pretrust]: Valid availability concern. If only `python3.10` exists without a `python3` shim, current discovery can skip despite a usable interpreter. I have merged main and retriggered checks; I will extend interpreter candidate coverage and add a direct test for versioned-only Python binaries.

File: scripts/cursor-worker-pretrust [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+  if "$python_candidate" -c 'import sys; assert sys.version_info[0] >= 3; import json' >/dev/null 2>&1; then
```
[the remark this answers is not in the document]
Comment (baksohyeon) [reply scripts/cursor-worker-pretrust]: Agreed. The probe should assert minimum runtime capability for the embedded script, not only major-version + json import. I have merged main and retriggered checks; I will tighten the probe to a minimum compatible version/capability and add regression coverage.

File: master-ops/docs/charter/04-worker-routing-review.md [hunk 15]
```diff
@@ -18,7 +18,7 @@ The installation-specific worker tier mapping lives in `model-tier-policy.json`,
 
 ### Worker Launch Approval Posture
```
[the remark this answers is not in the document]
Comment (baksohyeon) [reply master-ops/docs/charter/04-worker-routing-review.md]: Thanks — valid style issue. I have merged main and retriggered checks; I will change this heading level to keep strict Markdown hierarchy and clear the MD001 warning.

File: scripts/cursor-worker-pretrust (hunk 9 above)
[the remark this answers is not in the document]
Comment (baksohyeon) [reply scripts/cursor-worker-pretrust]: Fixed in 0dcfef6. The embedded writer now opens `projects_dir` and the project directory with `O_DIRECTORY|O_NOFOLLOW`, reads/writes the marker via `openat`-style no-follow FDs, and replaces through an exclusive temp file in the same directory. Symlink preflight remains as defense in depth; the write path no longer depends only on a shell check-then-Path.write race window. Regression coverage: existing symlink tests plus non-regular and trailing-slash cases.

File: scripts/cursor-worker-pretrust (hunk 10 above)
[the remark this answers is not in the document]
Comment (baksohyeon) [reply scripts/cursor-worker-pretrust:52]: Fixed in 0dcfef6. `check_no_symlink_ancestors` now strips trailing separators before each `-L` test so `--projects-dir /path/to/link/` still rejects the symlinked leaf. Measured before fix: trailing-slash projects-dir wrote the marker into the symlink target (exit 0); after fix: `ERROR trust marker path must not use symlinks` (exit 2). Added `test_symlinked_projects_dir_with_trailing_slash_is_rejected`.

File: scripts/cursor-worker-pretrust (hunk 11 above)
[the remark this answers is not in the document]
Comment (baksohyeon) [reply scripts/cursor-worker-pretrust]: Fixed in 0dcfef6. Before parse/update, the marker is opened with `O_NOFOLLOW` and `fstat` must report a regular file; directories and other non-regular nodes abort with a stable `trust marker is not a regular file` error (exit 2) instead of an uncaught `IsADirectoryError`. Added `test_non_regular_marker_is_rejected`.

File: tests/test_cursor_pretrust.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (baksohyeon) [reply tests/test_cursor_pretrust.py:49]: Fixed in 0dcfef6. Tests now resolve fixture roots via `tmp_path.resolve()` so macOS `/var` → `/private/var` no longer places symlink ancestors on the projects path under test. Measured: default pytest `tmp_path` sits under `/var/folders` where `/var` is a symlink; resolved fixtures keep happy-path coverage without weakening production symlink rejection.

File: scripts/cursor-worker-pretrust (hunk 13 above)
[the remark this answers is not in the document]
Comment (baksohyeon) [reply scripts/cursor-worker-pretrust]: Fixed in 0dcfef6. Interpreter discovery now walks `python3`, `python3.14`…`python3.6`, then `python`, so a host with only `python3.10` is accepted. Added `test_only_python310_candidate_is_accepted`.

File: scripts/cursor-worker-pretrust (hunk 14 above)
[the remark this answers is not in the document]
Comment (baksohyeon) [reply scripts/cursor-worker-pretrust]: Fixed in 0dcfef6. The probe now requires `sys.version_info >= (3, 6)`, `import json`, and a successful `datetime.isoformat(timespec="milliseconds")` call—matching the embedded worker—before selecting an interpreter. Unsuitable candidates fall through to the documented `Summary: skipped` path. Coverage: `test_python_candidate_below_minimum_is_rejected`.

File: master-ops/docs/charter/04-worker-routing-review.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (baksohyeon) [reply master-ops/docs/charter/04-worker-routing-review.md]: Fixed in 0dcfef6. Changed `### Worker Launch Approval Posture` to `## Worker Launch Approval Posture` in `master-ops/docs/charter/04-worker-routing-review.md` so the heading increments by one level from the document H1 and clears MD001.

File: docs/public/reference.md [hunk 16]
```diff
@@ -13,6 +13,7 @@ Use this page for command discovery. The lifecycle and delegation guides explain
...
 | `scripts/codex-worker-pretrust` | `codex-worker-pretrust` | Mark a worktree as trusted in every Orca-managed Codex account config, so a dispatched worker does not stop on a trust prompt. | Positional absolute worktree path; `--accounts-dir`. Finds its own TOML-capable interpreter (`python3`, newer versioned names, `python`) and, when none has `tomllib`, skips loudly without touching any config. |
+| `scripts/cursor-worker-pretrust` | `cursor-worker-pretrust` | Mark a worktree as trusted in Cursor Agent's measured trust storage, so a dispatched worker does not stop on the workspace-trust prompt. | Positional absolute worktree path; `--projects-dir`. Writes the measured `.workspace-trusted` marker under `~/.cursor/projects/<project-key>/` after validating existing JSON; when no Python 3 JSON interpreter is available, skips loudly without touching any marker. Follow-up measurement on `cursor-agent 2026.07.23-e383d2b` with project-local `.cursor/hooks.json` found no second startup trust gate (`hooks.state`/`trusted_hash`) to pre-seed. |
```
[the remark this answers is not in the document]
Comment (baksohyeon) [reply docs/public/reference.md]: Fixed in 0dcfef6. Removed the build-specific second-gate measurement note from the public reference Key options cell. That measurement remains in the runbook (`master-ops/docs/charter/04-worker-routing-review.md` and `05-dispatch-gate.md`) and both changelogs; the reference row now stays on `--projects-dir` / interpreter skip behavior only.
